spec/services/migration/migrators/user_spec.rb

```diff
@@ -79,6 +93,120 @@ def setup_failure_state
+          user = User.find_by(uid: ecf_user.get_an_identity_id)
+          expect(user.ecf_id).to eq(ecf_user.id)
+          expect(user.full_name).to eq("New Name")
+        end
+      end
+
+      context "when one user exists with ecf_id only" do
+        it "updates the user" do
+          ecf_user = create(:ecf_migration_user, :npq, get_an_identity_id: SecureRandom.uuid, full_name: "New Name")
+          create(:user, email: ecf_user.email, ecf_id: ecf_user.id, uid: SecureRandom.uuid, full_name: "Old Name")
+
+          expect(User.where(ecf_id: ecf_user.id).count).to eq(1)
+          expect(User.where(uid: ecf_user.get_an_identity_id).count).to eq(0)
+
+          instance.call
+
+          expect(User.where(ecf_id: ecf_user.id).count).to eq(1)
+          user = User.find_by(uid: ecf_user.get_an_identity_id)
+          expect(user.ecf_id).to eq(ecf_user.id)
+          expect(user.full_name).to eq("New Name")
+        end
+      end
+
+      context "when ecf_id and get_an_identity_id both return users and are different" do
+        it "raises error" do
+          ecf_user = create(:ecf_migration_user, :npq, email: "email-1@example.com", get_an_identity_id: SecureRandom.uuid)
+          create(:user, ecf_id: ecf_user.id, email: "email-1@example.com")
+          create(:user, uid: ecf_user.get_an_identity_id, email: "email-2@example.com")
+
+          instance.call
+          expect(failure_manager).to have_received(:record_failure).with(ecf_user, /ecf_user.id and ecf_user.get_an_identity_id both return User records, but they are different/)
+        end
+      end
+
+      context "when NPQ user found with ecf_user.get_an_identity_id only" do
+        let(:ecf_user) { create(:ecf_migration_user, :npq, email: "email-1@example.com", get_an_identity_id: SecureRandom.uuid, full_name: "New Name") }
+        let(:user) do
+          create(:user, email: ecf_user.email, ecf_id: SecureRandom.uuid, uid: ecf_user.get_an_identity_id, full_name: "Old Name")
+        end
+
+        context "when NPQ user.ecf_id is set but no ecf user found" do
+          it "updates user.ecf_id with ecf_user.id" do
+            expect(Migration::Ecf::User.find_by(id: user.ecf_id)).to be_nil
+
+            instance.call
+            expect(user.reload.ecf_id).to eq(ecf_user.id)
+            expect(user.full_name).to eq("New Name")
+          end
+        end
+
+        context "when NPQ user.ecf_id links to a different ecf_user" do
+          context "when linked ecf_user is an orphan" do
+            it "updates user.ecf_id with ecf_user.id" do
+              orphaned_ecf_user = create(:ecf_migration_user, id: user.ecf_id)
+              expect(orphaned_ecf_user.npq_applications).to be_empty
+
+              instance.call
+              expect(user.reload.ecf_id).to eq(ecf_user.id)
+              expect(user.full_name).to eq("New Name")
+            end
+          end
+
+          context "when linked ecf_user is not an orphan" do
+            let(:failure_manager) do
```

Comment:
We mock this already in the shared context, so you can test failures similar to this:

https://github.com/DFE-Digital/npq-registration/blob/main/spec/services/migration/migrators/application_state_spec.rb#L36

Reply:
We need to check for multiple failure messages. The mock doesn't work for that.

Reply:
I'm sure it does 😅 

```
expect(failure_manager).to have_received(:record_failure).once.with(ecf_user, "Validation failed: User found with ecf_user.get_an_identity_id, but its user.ecf_id linked to another ecf_user that is not an orphan")
expect(failure_manager).to have_received(:record_failure).once.with(non_orphaned_ecf_user, "Validation failed: Participant identity email from ECF does not match existing user email in NPQ")
```

Reply:
Yes thats what i needed lol Updated code now.